sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,467 @@
+
+template <typename ReturnT, typename ... ArgsT, ReturnT(&f) (ArgsT...),
+  class propertiesT>
+struct is_property_key_of<response_capacity_key,
+  task_sequence<f, propertiesT>> : std::true_type {};
+
+} // namespace experimental
+} // namespace intel
+} // namespace ext
+} // namespace sycl
+----
+
+--
+[options="header"]
+|===
+| Property | Description
+| balanced | The `balanced` property is a guarantee to the SYCL device compiler
+that the `task_sequence` object will call exactly the same number of `async` s
+and `get` s over the object's lifetime (i.e. before the `task_sequence`
+destructor is invoked).
+
+| invocation_capacity | The `async` invocations are guaranteed to not block
+as long as the number of outstanding invocations are less than or equal to
+`invocation_capacity`. An outstanding invocation is a call to `async` whose
+corresponding result has not yet been retrieved by a call to `get`.
+
+A default value is chosen by the implementation.
+
+| response_capacity | Given a response capacity of `N`, the user is guaranteed
+that at least one outstanding `async` call will make progress as long as the
+number of outstanding invocations is less than or equal to `N`.
+
+A default value is chosen by the implementation.
+|===
+--
+
+=== Compatibility with FPGA Kernel Interface Properties
+
+A `task_sequence` may be declared with the following FPGA Kernel Interface
+properties:
+
+ - `pipelined`
+ - `stall_enable_clusters`
+
+These are described in the link:../sycl_ext_oneapi_kernel_properties.asciidoc[
+sycl_ext_intel_fpga_kernel_interface_properties] document.
+
+Normally these properties are applicable only to kernels however this extension
+supports applying the properties to task sequences.
+
+== Forward Progress Guarantees and Execution Model
+
+C{plus}{plus} defines a framework for describing the
+http://eel.is/c++draft/intro.progress[forward progress] of individual threads
+with respect to one another in a multi-threaded program.
+
+Applying this framework to `task_sequence`, `async` tasks belonging to the same
+`task_sequence` object provide a http://eel.is/c++draft/intro.progress#11[
+weakly parallel forward progress guarantee]. That is, tasks belonging to a
+`task_sequence` object are not guaranteed to make forward progress with respect
+to each other.
+
+Tasks belonging to different `task_sequence` objects provide a
+http://eel.is/c++draft/intro.progress#9[parallel forward progress guarantee]
+with respect to each other. That is, tasks belonging to different
+`task_sequence` objects are guaranteed to make forward progress with respect to
+each other once they have been initiated.
+
+Tasks belonging to a `task_sequence` object provide a
+http://eel.is/c++draft/intro.progress#9[parallel forward progress guarantee]
+with respect to their caller. That is, a task sequence's tasks are guaranteed
+to make forward progress with respect to their caller (that invoked `async`).
```

Comment:
I think the phrasing "with respect to" is problematic here, because that isn't aligned with my understanding of how C++ defines forward progress.

If tasks have weakly parallel forward progress guarantees, then they have no guarantees.  You can't say that they also have parallel forward progress guarantees.  I think what you've written here says that if you only have one `task_sequence` object, then the tasks in that sequence are never guaranteed to execute until a work-item calls `get()` and blocks upon their result, and an implementation can switch between tasks in the sequence at any time.  But if you have two `task_sequence` objects, once a task starts executing it will continue to do so until it terminates.  I don't think that's what you mean.

I think what you want to say is something closer to:

- Each `task_sequence` is associated with one thread of execution that provides parallel forward progress guarantees.
- Each call to `async` creates a new task, associated with a thread of execution that provides weakly parallel forward progress guarantees.
- The thread associated with the `task_sequence` blocks with forward progress guarantee delegation upon the tasks in the task sequence.
- When `get()` is called, the calling thread of execution (i.e. work-item) blocks with forward progress guarantee delegation upon the thread of execution associated with the `task_sequence`.

I still don't think this is exactly right, though, because an implementation still wouldn't be required to ensure any tasks make progress until you call `get()`.  The language about `response_capacity` seems to suggest you want to guarantee some of these tasks will start executing even without a call to `get()`... I'm not 100% sure, but I think you might need a `task_sequence` to be associated with `response_capacity` threads of execution providing concurrent forward progress guarantees in that case.

Reply:
Hey John, thank you for taking the time to write this detailed response. After much discussion and thinking, we have re-written this section. The updates are being discussed on this [SPIRV extension PR](https://github.com/intel-innersource/documentation.xpu.architecture.spirv-extension-drafts/pull/107). Could you please review the new description there?
It will address most of your concerns above if not all. This section about progress guarantees exists in both the SPIRV extension and this SYCL extension. It's just receiving more comments there. Once the review concludes, I'll figure out whether to paste that result here or to link one extension to the other to avoid duplication.

Reply:
I have used the material from the SPIRV extension PR and written up a section about Forward Progress Guarantees and Execution Model, please take a look!